test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -410,27 +410,34 @@ TEST_VM(os, jio_snprintf) {
 #endif
 
 TEST_VM(os, realpath) {
-  static const char* path = "/1234567890123456789";
+  /* POSIX requires that the file exists, Windows doesn't */
+  static const char* nosuchpath = "/1234567890123456789";
+  static const char* tmppath = "/tmp";
 
-	char buffer[MAX_PATH];
+  char buffer[MAX_PATH];
 
   errno = 0;
-	const char* returnedBuffer = os::realpath(path, buffer, 10);
-  EXPECT_TRUE(errno == ENAMETOOLONG);
+  const char* returnedBuffer = os::realpath(nosuchpath, buffer, sizeof(nosuchpath) - 2);
+  /* Returns ENOENT on Linux, ENAMETOOLONG on Windows */
   EXPECT_TRUE(returnedBuffer == nullptr);
+  EXPECT_TRUE(errno == ENAMETOOLONG || errno == ENOENT);
+
+  errno = 0;
+  returnedBuffer = os::realpath(nosuchpath, buffer, sizeof(nosuchpath) + 3);
+  /* Returns ENOENT on Linux, 0 on Windows */
+  EXPECT_TRUE(errno == 0 || errno == ENOENT);
+  EXPECT_TRUE(returnedBuffer == nullptr || returnedBuffer == buffer);
 
-	errno = 0;
-	returnedBuffer = os::realpath(path, buffer, MAX_PATH);
-  EXPECT_TRUE(errno == 0);
+  errno = 0;
+  returnedBuffer = os::realpath(tmppath, buffer, MAX_PATH);
   EXPECT_TRUE(returnedBuffer == buffer);
```

Comment:
Should we also do `EXPECT_TRUE(errno == 0);` ? Here and below.

Reply:
This is interesting!  I found that on Linux, errno _was not zero_!  The specifications for POSIX realpath say
`RETURN VALUE
Upon successful completion, realpath() shall return a pointer to the resolved name. Otherwise, realpath() shall return a null pointer and set errno to indicate the error, and the contents of the buffer pointed to by resolved_name are undefined.`
Nowhere does it say errno is unchanged if successful.

`  errno = 0;
  ::printf("before ::realpath(\"/tmp\",nullptr) errno=%d\n", errno);
  char* p = ::realpath("/tmp", nullptr);
  ::printf("after ::realpath p=%s errno=%d\n", p, errno);`

outputs:
`before ::realpath("/tmp",nullptr) errno=0
after ::realpath /tmp p=/tmp errno=22
`
With behaviour like this, one can see why OpenJDK wraps ::realpath()...

Reply:
Right I forgot about this. The Posix spec even states:
>  No function in this volume of POSIX.1-2017 shall set errno to 0.
so reading errno is only valid after calling a function that sets errno on error, and which returns a value that says there was an error. Windows GetLastError is the same. Even setting it to zero before the call does not guarantee the call doesn't modify it even if successful.